tests/e2e/tidbcluster/tidbcluster.go

```diff
@@ -1966,6 +1937,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
+
+					ginkgo.By(fmt.Sprintf("Scale in %s to %d replicas", comp, replicasSmall))
+					err := controller.GuaranteedUpdate(genericCli, tc, func() error {
+						switch comp {
+						case v1alpha1.PDMemberType:
+							tc.Spec.PD.Replicas = replicasSmall
+						case v1alpha1.TiKVMemberType:
+							tc.Spec.TiKV.Replicas = replicasSmall
+						case v1alpha1.TiDBMemberType:
+							tc.Spec.TiDB.Replicas = replicasSmall
+						}
+						return nil
+					})
+					framework.ExpectNoError(err, "failed to scale in %s for TidbCluster %s/%s", comp, ns, tc.Name)
+
+					ginkgo.By(fmt.Sprintf("Wait for %s to be in ScalePhase", comp))
+					utiltc.MustWaitForComponentPhase(cli, tc, comp, v1alpha1.ScalePhase, 3*time.Minute, 10*time.Second)
+
+					ginkgo.By(fmt.Sprintf("Upgrade %s version concurrently", comp))
+					err = controller.GuaranteedUpdate(genericCli, tc, func() error {
+						switch comp {
+						case v1alpha1.PDMemberType:
+							tc.Spec.PD.Version = pointer.StringPtr(utilimage.TiDBV4)
+						case v1alpha1.TiKVMemberType:
+							tc.Spec.TiKV.Version = pointer.StringPtr(utilimage.TiDBV4)
+						case v1alpha1.TiDBMemberType:
+							tc.Spec.TiDB.Version = pointer.StringPtr(utilimage.TiDBV4)
+						}
+						return nil
+					})
+					framework.ExpectNoError(err, "failed to scale in %s for TidbCluster %s/%s", comp, ns, tc.Name)
+
+					ginkgo.By("Wait for tc ready")
+					err = oa.WaitForTidbClusterReady(tc, 10*time.Minute, 10*time.Second)
+					framework.ExpectNoError(err, "failed to wait for TidbCluster %s/%s ready after scale in %s", ns, tc.Name, comp)
+
+					ginkgo.By(fmt.Sprintf("Check %s Pods", comp))
+					var labelSelector string
+					switch comp {
+					case v1alpha1.PDMemberType:
+						labelSelector = labels.SelectorFromSet(label.New().Instance(tc.Name).PD().Labels()).String()
+					case v1alpha1.TiKVMemberType:
+						labelSelector = labels.SelectorFromSet(label.New().Instance(tc.Name).TiKV().Labels()).String()
+					case v1alpha1.TiDBMemberType:
+						labelSelector = labels.SelectorFromSet(label.New().Instance(tc.Name).TiDB().Labels()).String()
+					}
+					listOptions := metav1.ListOptions{
+						LabelSelector: labelSelector,
+					}
+					pods, err := c.CoreV1().Pods(ns).List(listOptions)
+					framework.ExpectNoError(err, "failed to list %s Pods with options: %+v", comp, listOptions)
+					framework.ExpectEqual(len(pods.Items), int(replicasSmall), "there should be %d %s Pods", replicasSmall, comp)
+					for _, pod := range pods.Items {
+						// some pods may have multiple containers
+						wrongImage := true
+						for _, c := range pod.Spec.Containers {
+							log.Logf("container image: %s", c.Image)
+							if fmt.Sprintf("pingcap/%s:%s", comp, utilimage.TiDBV4) == c.Image {
+								wrongImage = false
+								break
+							}
+						}
+						if wrongImage {
+							log.Failf("%s Pod has wrong image, expected %s", comp, utilimage.TiDBV4)
```

Comment:
How about using `framework.Failf` instead?

Reply:
They're actually different wrappers for `FailfWithOffset(1, format, args...)`